src/backend/internal/controller/user/update_user_profile.go

```diff
@@ -0,0 +1,117 @@
+package user
+
+import (
+	"errors"
+	"math/rand"
+	"net/http"
+
+	"github.com/Roongkun/software-eng-ii/internal/controller/user/fieldvalidate"
+	"github.com/Roongkun/software-eng-ii/internal/controller/util"
+	"github.com/Roongkun/software-eng-ii/internal/model"
+	"github.com/gin-gonic/gin"
+	"golang.org/x/crypto/bcrypt"
+)
+
+func (r *Resolver) UpdateUserProfile(c *gin.Context) {
+
+	user := c.MustGet("user")
+	userObj, ok := user.(model.User)
+	if !ok {
+		err := errors.New("Invalid user type in context")
+		util.Raise500Error(c, err)
+		return
+	}
+
+	updatingUserInput := model.UserUpdateInput{}
+	if err := c.BindJSON(&updatingUserInput); err != nil {
+		util.Raise400Error(c, "unable to bind request body with json model, please recheck")
+		return
+	}
+
+	if fieldErrs := fieldvalidate.UpdateUser(updatingUserInput); len(fieldErrs) > 0 {
+		c.JSON(http.StatusBadRequest, gin.H{
+			"status": "failed",
+			"error":  util.JSONErrs(fieldErrs),
+		})
+		c.Abort()
+		return
+	}
+
+	if updatingUserInput.Username != nil {
+		existingUser, err := r.UserUsecase.FindOneByUsername(c, *updatingUserInput.Username)
+		if err == nil && existingUser.Id != userObj.Id {
+			c.JSON(http.StatusConflict, gin.H{
+				"status": "failed",
+				"error":  "Username already exists.",
+			})
+			return
+		} else if err != nil && err.Error() != "sql: no rows in result set" {
```

Comment:
Hard-coding like this could be dangerous when third-party libraries, such as `bun-uptrace` or `postgresql` in our case, make some updates. This could break our application's logic.

There are many approaches to handle this, here I would suggest one:

Consider changing your function in the `usecase` package to 
```CheckUsernameAlreadyBeenUsed(ctx context.Context, username string, proposedUserId uuid.UUID) (bool, error)```

Afterwards, your SQL statement could be 

```exist, err := u.db.NewSelect().Model(&user).Where("username = ? AND id != ?", username, proposedUserId).Exists(ctx)```

Then, return `exist` and `err` and proceed with your logic.

This way the code here would be cleaner, which it should be as this package serves as a high level controller.

If you have any other suggestions, please do not hesitate to discuss with me!